core/src/error.rs

```diff
@@ -293,6 +293,15 @@ pub enum TypecheckError {
         RichTerm, /* the inferred flat type */
         TermPos,
     ),
+    /// Types can appear in term position, like the `Number` in `let c = Number in 5 | c`.
+    /// These types are treated as contracts, but they are not allowed to contain flat types.
+    /// For example, `let c = { foo : 5 } in ...` is disallowed.
```

Comment:
I think this comment is a tad confusing, because it never alludes to the fact that this restriction is only valid for statically typed code. In statically typed code, we also reject valid things, such as `MyContract -> Number`.

Conversely, we don't reject `let c = { foo : 5}` inside dynamically typed code. I think we could say something along the lines : 

> within statically typed code, the typechecker must reject terms containing nonsensical contracts such as `let C = { foo : 5} in ({ foo = 5 } | C)`, which will fail at runtime. The typechecker is currently quite conservative and simply forbids to store any custom contract (flat type) in type that appears in term position. Note that this restriction doesn't apply to annotations (so `{foo = 5} | {foo : 5}` is accepted), which are type positions, and aren't considered part of the statically typed block per se.

Might need some rewording.